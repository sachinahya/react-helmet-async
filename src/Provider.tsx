import React, { FC, ReactNode } from 'react';
import { HelmetStateClient } from './HelmetState';

export const Context = React.createContext<HelmetStateClient>({} as HelmetStateClient);

export interface HelmetProviderProps {
  state: HelmetStateClient;
  children?: ReactNode;
}

export const HelmetProvider: FC<HelmetProviderProps> = ({ state, children }) => {
  return <Context.Provider value={state}>{children}</Context.Provider>;
};
