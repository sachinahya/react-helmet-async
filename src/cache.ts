import { HelmetProps } from './state';

export interface HelmetCache {
  update(instance: unknown, props: HelmetProps): void;
  remove(instance: unknown): void;
}
