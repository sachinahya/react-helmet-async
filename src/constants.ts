import { LiteralUnion } from 'type-fest';

export const TAG_PROPERTIES = {
  CHARSET: 'charset',
  CSS_TEXT: 'cssText',
  HREF: 'href',
  HTTPEQUIV: 'http-equiv',
  INNER_HTML: 'innerHTML',
  ITEM_PROP: 'itemprop',
  NAME: 'name',
  PROPERTY: 'property',
  REL: 'rel',
  SRC: 'src',
} as const;

export const ATTRIBUTE_NAMES = {
  BODY: 'bodyAttributes',
  HTML: 'htmlAttributes',
  TITLE: 'titleAttributes',
} as const;

export const TAG_NAMES = {
  BASE: 'base',
  BODY: 'body',
  HEAD: 'head',
  HTML: 'html',
  LINK: 'link',
  META: 'meta',
  NOSCRIPT: 'noscript',
  SCRIPT: 'script',
  STYLE: 'style',
  TITLE: 'title',
  FRAGMENT: 'Symbol(react.fragment)',
} as const;

export const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['generator', 'robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
} as const;

export const VALID_TAG_NAMES = Object.values(TAG_NAMES);

export const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
} as const;

const swapKeysAndValues = <T extends Record<string | number | symbol, string | number | symbol>>(
  obj: T
): { [K in keyof T as T[K]]: K } => {
  return Object.fromEntries(Object.entries(obj).map(([key, value]) => [value, key])) as {
    [K in keyof T as T[K]]: K;
  };
};

export const HTML_TAG_MAP = swapKeysAndValues(REACT_TAG_MAP);

export const HELMET_ATTRIBUTE = 'data-rh';
