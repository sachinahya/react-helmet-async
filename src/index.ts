export { Helmet } from './Helmet';
export { type HelmetStateClient as HelmetState } from './HelmetState';
