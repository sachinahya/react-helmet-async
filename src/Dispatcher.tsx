import { Component } from 'react';
import shallowEqual from 'shallowequal';
import handleStateChangeOnClient from './client';
import mapStateOnServer from './server';
import { reducePropsToState } from './utils';
import Provider, { HelmetContext } from './Provider';
import { HelmetProps } from './index';

export interface DispatcherProps extends HelmetProps {
  context: HelmetContext;
}

export default class Dispatcher extends Component<DispatcherProps> {
  rendered = false;

  shouldComponentUpdate(nextProps) {
    return !shallowEqual(nextProps, this.props);
  }

  componentDidUpdate() {
    this.emitChange();
  }

  componentWillUnmount() {
    const { helmetInstances } = this.props.context;
    helmetInstances.remove(this);
    this.emitChange();
  }

  emitChange() {
    const { helmetInstances, setHelmet } = this.props.context;
    let serverState = null;
    const propsList = helmetInstances.get().map(instance => {
      const props = { ...instance.props };
      delete props.context;
      return props;
    });
    const state = reducePropsToState(propsList);
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      serverState = mapStateOnServer(state);
    }
    setHelmet(serverState);
  }

  // componentWillMount will be deprecated
  // for SSR, initialize on first render
  // constructor is also unsafe in StrictMode
  init() {
    if (this.rendered) {
      return;
    }

    this.rendered = true;

    const { helmetInstances } = this.props.context;
    helmetInstances.add(this);
    this.emitChange();
  }

  render() {
    this.init();

    return null;
  }
}
