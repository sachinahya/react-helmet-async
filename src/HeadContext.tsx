import { FC, ReactNode, createContext } from 'react';
import { HeadCache } from './cache';

export const HeadContext = createContext<HeadCache | undefined>(undefined);

export interface HeadProviderProps {
  state: HeadCache;
  children?: ReactNode;
}

export const HeadProvider: FC<HeadProviderProps> = ({ state, children }) => {
  return <HeadContext.Provider value={state}>{children}</HeadContext.Provider>;
};
