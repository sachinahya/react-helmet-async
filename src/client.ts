import { HELMET_ATTRIBUTE, TAG_NAMES, TAG_PROPERTIES } from './constants';
import { ReducedState, flattenArray } from './utils';

const updateTags = <
  T extends Omit<typeof TAG_NAMES, 'FRAGMENT'>[keyof Omit<typeof TAG_NAMES, 'FRAGMENT'>],
>(
  type: T,
  tags: ReducedState[
    | 'baseTag'
    | 'linkTags'
    | 'metaTags'
    | 'noscriptTags'
    | 'scriptTags'
    | 'styleTags']
) => {
  type Tag = (typeof tags)[number];

  const headElement = document.head || document.querySelector(TAG_NAMES.HEAD);
  const tagNodes = headElement.querySelectorAll<HTMLElementTagNameMap[T]>(
    `${type}[${HELMET_ATTRIBUTE}]`
  );
  const oldTags: HTMLElementTagNameMap[T][] = [].slice.call(tagNodes);
  const newTags: HTMLElementTagNameMap[T][] = [];
  let indexToDelete: number;

  if (tags && tags.length) {
    tags.forEach((tag: Tag) => {
      const newElement = document.createElement<T>(type);

      // eslint-disable-next-line
      for (const attribute in tag) {
        if (Object.prototype.hasOwnProperty.call(tag, attribute)) {
          if (attribute === TAG_PROPERTIES.INNER_HTML) {
            newElement.innerHTML = tag.innerHTML;
          } else if (attribute === TAG_PROPERTIES.CSS_TEXT) {
            if (newElement.styleSheet) {
              newElement.styleSheet.cssText = tag.cssText;
            } else {
              newElement.appendChild(document.createTextNode(tag.cssText));
            }
          } else {
            const value = typeof tag[attribute] === 'undefined' ? '' : tag[attribute];
            newElement.setAttribute(attribute, value);
          }
        }
      }

      newElement.setAttribute(HELMET_ATTRIBUTE, 'true');

      // Remove a duplicate tag from domTagstoRemove, so it isn't cleared.
      if (
        oldTags.some((existingTag, index) => {
          indexToDelete = index;
          return newElement.isEqualNode(existingTag);
        })
      ) {
        oldTags.splice(indexToDelete, 1);
      } else {
        newTags.push(newElement);
      }
    });
  }

  oldTags.forEach(tag => tag.parentNode.removeChild(tag));
  newTags.forEach(tag => headElement.appendChild(tag));

  return {
    oldTags,
    newTags,
  };
};

const updateAttributes = <T extends keyof ReducedState>(
  tagName: (typeof TAG_NAMES)[keyof typeof TAG_NAMES],
  attributes: ReducedState[T]
): void => {
  const elementTag = document.getElementsByTagName(tagName)[0];

  if (!elementTag) {
    return;
  }

  const helmetAttributeString = elementTag.getAttribute(HELMET_ATTRIBUTE);
  const helmetAttributes = helmetAttributeString ? helmetAttributeString.split(',') : [];
  const attributesToRemove = [].concat(helmetAttributes);
  const attributeKeys = Object.keys(attributes);

  for (let i = 0; i < attributeKeys.length; i += 1) {
    const attribute = attributeKeys[i];
    const value = attributes[attribute] || '';

    if (elementTag.getAttribute(attribute) !== value) {
      elementTag.setAttribute(attribute, value);
    }

    if (helmetAttributes.indexOf(attribute) === -1) {
      helmetAttributes.push(attribute);
    }

    const indexToSave = attributesToRemove.indexOf(attribute);
    if (indexToSave !== -1) {
      attributesToRemove.splice(indexToSave, 1);
    }
  }

  for (let i = attributesToRemove.length - 1; i >= 0; i -= 1) {
    elementTag.removeAttribute(attributesToRemove[i]);
  }

  if (helmetAttributes.length === attributesToRemove.length) {
    elementTag.removeAttribute(HELMET_ATTRIBUTE);
  } else if (elementTag.getAttribute(HELMET_ATTRIBUTE) !== attributeKeys.join(',')) {
    elementTag.setAttribute(HELMET_ATTRIBUTE, attributeKeys.join(','));
  }
};

const updateTitle = (title: string, attributes: ReducedState['titleAttributes']) => {
  if (typeof title !== 'undefined' && document.title !== title) {
    document.title = flattenArray(title);
  }

  updateAttributes(TAG_NAMES.TITLE, attributes);
};

const commitTagChanges = (newState: ReducedState, cb?: () => void) => {
  const {
    baseTag,
    bodyAttributes,
    htmlAttributes,
    linkTags,
    metaTags,
    noscriptTags,
    onChangeClientState,
    scriptTags,
    styleTags,
    title,
    titleAttributes,
  } = newState;
  updateAttributes(TAG_NAMES.BODY, bodyAttributes);
  updateAttributes(TAG_NAMES.HTML, htmlAttributes);

  updateTitle(title, titleAttributes);

  const tagUpdates = {
    baseTag: updateTags(TAG_NAMES.BASE, baseTag),
    linkTags: updateTags(TAG_NAMES.LINK, linkTags),
    metaTags: updateTags(TAG_NAMES.META, metaTags),
    noscriptTags: updateTags(TAG_NAMES.NOSCRIPT, noscriptTags),
    scriptTags: updateTags(TAG_NAMES.SCRIPT, scriptTags),
    styleTags: updateTags(TAG_NAMES.STYLE, styleTags),
  };

  const addedTags = {};
  const removedTags = {};

  Object.keys(tagUpdates).forEach(tagType => {
    const { newTags, oldTags } = tagUpdates[tagType];

    if (newTags.length) {
      addedTags[tagType] = newTags;
    }
    if (oldTags.length) {
      removedTags[tagType] = tagUpdates[tagType].oldTags;
    }
  });

  if (cb) {
    cb();
  }

  onChangeClientState(newState, addedTags, removedTags);
};

// eslint-disable-next-line
let _helmetCallback = null;

const handleStateChangeOnClient = (newState: ReducedState) => {
  if (_helmetCallback) {
    cancelAnimationFrame(_helmetCallback);
  }

  if (newState.defer) {
    _helmetCallback = requestAnimationFrame(() => {
      commitTagChanges(newState, () => {
        _helmetCallback = null;
      });
    });
  } else {
    commitTagChanges(newState);
    _helmetCallback = null;
  }
};

export default handleStateChangeOnClient;
