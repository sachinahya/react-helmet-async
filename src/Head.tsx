import {
  Children,
  Component,
  ContextType,
  PropsWithChildren,
  ReactElement,
  ReactNode,
  isValidElement,
} from 'react';
import { isFragment } from 'react-is';
import fastCompare from 'react-fast-compare';
import invariant from 'tiny-invariant';
import { HeadContext } from './HeadContext';
import { NON_SELF_CLOSING_TAGS, TAG_NAMES, VALID_TAG_NAMES } from './constants';
import { HeadProps } from './state';

export interface HeadComponentProps {
  children?: ReactNode;
}

function flattenArrayTypeChildren(
  arrayTypeChildren: Record<string, unknown[]>,
  child: ReactElement<PropsWithChildren>
): void {
  if (typeof child.type !== 'string') {
    return;
  }

  const newChildProps = { ...child.props };

  if (newChildProps.children) {
    const isSelfClosing = !NON_SELF_CLOSING_TAGS.includes(child.type);

    if (isSelfClosing) {
      throw new Error(
        `<${child.type} /> elements are self-closing and can not contain children. Refer to our API for more information.`
      );
    }
  }

  (arrayTypeChildren[child.type] ??= []).push(newChildProps);
}

function mapObjectTypeChildren(
  mappedProps: HeadProps,
  child: ReactElement<PropsWithChildren>
): HeadProps {
  const { children: nestedChildren, ...newChildProps } = child.props;

  switch (child.type) {
    case TAG_NAMES.TITLE: {
      // eslint-disable-next-line no-nested-ternary
      mappedProps.title = Array.isArray(nestedChildren)
        ? nestedChildren
        : nestedChildren
        ? String(nestedChildren)
        : undefined;

      mappedProps.titleAttributes = newChildProps;

      break;
    }

    case TAG_NAMES.BODY:
      mappedProps.bodyAttributes = newChildProps;
      break;

    case TAG_NAMES.HTML:
      mappedProps.htmlAttributes = newChildProps;
      break;

    default: {
      // eslint-disable-next-line no-lone-blocks
      if (typeof child.type === 'string') {
        (mappedProps as Record<string, unknown>)[child.type] = newChildProps;
      }

      break;
    }
  }

  return mappedProps;
}

function mapChildrenToProps(componentProps: Readonly<HeadComponentProps>): HeadProps {
  const arrayTypeChildren: Record<string, unknown[]> = {};

  const { children, ...mappedProps } = componentProps;

  Children.forEach(children as ReactElement<PropsWithChildren>[], child => {
    if (!isValidElement(child)) {
      return;
    }

    // isFragment incorrectly narrows to ReactElement.
    // A child can still be a ReactElement if it's not a fragment.
    if (isFragment(child) as boolean) {
      Object.assign(
        mappedProps,
        mapChildrenToProps({
          ...mappedProps,
          children: child.props.children,
        })
      );
    } else {
      invariant(
        VALID_TAG_NAMES.some(name => child.type === name),
        typeof child.type === 'function'
          ? `You may be attempting to nest <Head> components within each other, which is not allowed. Refer to our API for more information.`
          : `Only elements types ${VALID_TAG_NAMES.join(
              ', '
            )} are allowed. Head does not support rendering <${
              child.type
            }> elements. Refer to our API for more information.`
      );

      invariant(
        !child.props.children ||
          typeof child.props.children === 'string' ||
          (Array.isArray(child.props.children) &&
            !child.props.children.some(nestedChild => typeof nestedChild !== 'string')),
        `Head expects a string as a child of <${child.type}>. Did you forget to wrap your children in braces? ( <${child.type}>{\`\`}</${child.type}> ) Refer to our API for more information.`
      );

      switch (child.type) {
        case TAG_NAMES.BASE:
        case TAG_NAMES.LINK:
        case TAG_NAMES.META:
        case TAG_NAMES.NOSCRIPT:
        case TAG_NAMES.SCRIPT:
        case TAG_NAMES.STYLE:
          flattenArrayTypeChildren(arrayTypeChildren, child);
          break;

        default:
          mapObjectTypeChildren(mappedProps, child);
          break;
      }
    }
  });

  for (const [arrayChildName, arrayChild] of Object.entries(arrayTypeChildren)) {
    (mappedProps as Record<string, unknown[]>)[arrayChildName] = arrayChild;
  }

  return mappedProps;
}

const assertContext = (context: ContextType<typeof HeadContext>) => {
  invariant(context, 'Missing HeadProvider');
  return context;
};

export class Head extends Component<HeadComponentProps> {
  static override contextType = HeadContext;

  // @ts-expect-error
  override context!: ContextType<typeof HeadContext>;

  rendered: boolean = false;

  override shouldComponentUpdate(nextProps: HeadComponentProps) {
    return !fastCompare(this.props, nextProps);
  }

  override componentDidUpdate() {
    const mappedProps = mapChildrenToProps(this.props);
    assertContext(this.context).update(this, mappedProps);
  }

  override componentWillUnmount() {
    assertContext(this.context).remove(this);
  }

  override render() {
    if (!this.rendered) {
      this.rendered = true;

      const mappedProps = mapChildrenToProps(this.props);
      assertContext(this.context).update(this, mappedProps);
    }

    return null;
  }
}
