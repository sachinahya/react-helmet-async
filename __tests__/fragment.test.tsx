import React from 'react';
import ReactDOM from 'react-dom';
import { Helmet } from '../src';
import Provider from '../src/Provider';

Helmet.defaultProps.defer = false;

const mount = document.getElementById('mount');

const render = node => {
  ReactDOM.render(<Provider>{node}</Provider>, mount);
};

describe('fragments', () => {
  it('parses Fragments', () => {
    render(
      <Helmet>
        <>
          <title>Hello</title>
          <meta charSet="utf-8" />
        </>
      </Helmet>
    );

    expect(document.title).toMatchSnapshot();
  });

  it('parses nested Fragments', () => {
    render(
      <Helmet>
        <>
          <title>Foo</title>
          <>
            <title>Bar</title>
            <title>Baz</title>
          </>
        </>
      </Helmet>
    );

    expect(document.title).toMatchSnapshot();
  });
});
