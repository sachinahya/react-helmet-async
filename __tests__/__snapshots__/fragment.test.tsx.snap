// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragments parses Fragments 1`] = `"Hello"`;

exports[`fragments parses nested Fragments 1`] = `"Baz"`;
